libraries/chain/include/eosio/chain/fork_database.hpp

```diff
@@ -108,86 +109,111 @@ namespace eosio::chain {
    using fork_database_if_t     = fork_database_t<block_state_ptr>;
 
    /**
-    * Provides thread safety on fork_database_t and provide mechanism for opening the correct type
+    * Provides mechanism for opening the correct type
     * as well as switching from legacy (old dpos) to instant-finality.
     *
     * All methods assert until open() is closed.
     */
    class fork_database {
-      mutable std::recursive_mutex m;
       const std::filesystem::path data_dir;
-      std::variant<fork_database_t<block_state_legacy_ptr>, fork_database_t<block_state_ptr>> vforkdb;
+      std::atomic<bool> legacy = true;
+      std::unique_ptr<fork_database_legacy_t> fork_db_legacy;
+      std::unique_ptr<fork_database_if_t>     fork_db_if;
    public:
       explicit fork_database(const std::filesystem::path& data_dir);
       ~fork_database(); // close on destruction
 
+      // not thread safe, expected to be called from main thread before allowing concurrent access
       void open( validator_t& validator );
       void close();
 
+      // expected to be called from main thread, accesses chain_head
       void switch_from_legacy();
 
       // see fork_database_t::fetch_branch(forkdb->head()->id())
       std::vector<signed_block_ptr> fetch_branch_from_head();
 
       template <class R, class F>
       R apply(const F& f) {
-         std::lock_guard g(m);
-         if constexpr (std::is_same_v<void, R>)
-            std::visit([&](auto& forkdb) { f(forkdb); }, vforkdb);
-         else
-            return std::visit([&](auto& forkdb) -> R { return f(forkdb); }, vforkdb);
+         if constexpr (std::is_same_v<void, R>) {
+            if (legacy) {
+               f(*fork_db_legacy);
+            } else {
+               f(*fork_db_if);
+            }
+         } else {
+            if (legacy) {
+               return f(*fork_db_legacy);
+            } else {
+               return f(*fork_db_if);
+            }
+         }
```

Comment:
Personal preference, I prefer the more compact version (esp because we have many variations of `apply`), but fine either way.

```suggestion
         if constexpr (std::is_same_v<void, R>) {
            legacy ? f(*fork_db_legacy) : f(*fork_db_if);
         } else {
            return legacy ? f(*fork_db_legacy) :  return f(*fork_db_if);
         }
```

Reply:
I like compact code, but the current way allows for easier setting of breakpoints.

Reply:
Yes, that's true! I'm fine with leaving as is.